File: articles/many-body-physics-and-emergent-collective-behavior/python/occupation_statistics.py
"""
Bose and Fermi Occupation Statistics

This workflow compares ideal Fermi-Dirac and Bose-Einstein occupation
functions across energy and temperature.
"""

from pathlib import Path

import numpy as np
import pandas as pd


K_B_EV_K = 8.617333262e-5


def fermi_occupation(energy_ev: np.ndarray, mu_ev: float, temperature_k: float) -> np.ndarray:
    """
    Compute Fermi-Dirac occupation.
    """
    x = (energy_ev - mu_ev) / (K_B_EV_K * temperature_k)
    return 1.0 / (np.exp(x) + 1.0)


def bose_occupation(energy_ev: np.ndarray, mu_ev: float, temperature_k: float) -> np.ndarray:
    """
    Compute Bose-Einstein occupation where E > mu.
    """
    x = (energy_ev - mu_ev) / (K_B_EV_K * temperature_k)
    values = np.full_like(energy_ev, np.nan, dtype=float)
    valid = energy_ev > mu_ev
    values[valid] = 1.0 / (np.exp(x[valid]) - 1.0)
    return values


def run_case(case: pd.Series) -> pd.DataFrame:
    """
    Compute occupation table for one parameter case.
    """
    energies = np.linspace(
        float(case["energy_min_ev"]),
        float(case["energy_max_ev"]),
        int(case["n_points"]),
    )

    mu = float(case["chemical_potential_ev"])
    temperature = float(case["temperature_k"])

    return pd.DataFrame(
        {
            "case_id": case["case_id"],
            "temperature_k": temperature,
            "chemical_potential_ev": mu,
            "energy_ev": energies,
            "fermi_occupation": fermi_occupation(energies, mu, temperature),
            "bose_occupation": bose_occupation(energies, mu, temperature),
            "notes": case["notes"],
        }
    )


def main() -> None:
    """
    Generate occupation statistics tables.
    """
    article_dir = Path(__file__).resolve().parents[1]
    input_path = article_dir / "data" / "occupation_cases.csv"
    output_path = article_dir / "data" / "computed_occupation_statistics.csv"

    cases = pd.read_csv(input_path)
    output = pd.concat([run_case(case) for _, case in cases.iterrows()], ignore_index=True)

    output.to_csv(output_path, index=False)

    print("Occupation statistics sample:")
    print(output.groupby("case_id").head(5).round(8).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/many-body-physics-and-emergent-collective-behavior/r/occupation_statistics.R
# Bose and Fermi Occupation Statistics
#
# This workflow compares ideal occupation functions:
#
#   Fermi-Dirac:    f(E) = 1 / (exp((E - mu)/(k_B T)) + 1)
#   Bose-Einstein: n(E) = 1 / (exp((E - mu)/(k_B T)) - 1)

library(tidyverse)

article_dir <- "articles/many-body-physics-and-emergent-collective-behavior"

input_path <- file.path(article_dir, "data/occupation_cases.csv")
output_path <- file.path(article_dir, "data/computed_occupation_statistics_r.csv")
summary_path <- file.path(article_dir, "data/computed_occupation_statistics_summary_r.csv")

boltzmann_constant_ev_k <- 8.617333262e-5

cases <- read_csv(input_path, show_col_types = FALSE)

occupation_table <- cases %>%
  rowwise() %>%
  mutate(energy_ev = list(seq(energy_min_ev, energy_max_ev, length.out = n_points))) %>%
  unnest(energy_ev) %>%
  ungroup() %>%
  mutate(
    dimensionless_energy =
      (energy_ev - chemical_potential_ev) /
      (boltzmann_constant_ev_k * temperature_k),
    fermi_occupation =
      1 / (exp(dimensionless_energy) + 1),
    bose_occupation =
      if_else(
        energy_ev > chemical_potential_ev,
        1 / (exp(dimensionless_energy) - 1),
        NA_real_
      )
  )

summary_table <- occupation_table %>%
  group_by(case_id, temperature_k) %>%
  summarise(
    fermi_min = min(fermi_occupation, na.rm = TRUE),
    fermi_max = max(fermi_occupation, na.rm = TRUE),
    bose_max_finite = max(bose_occupation, na.rm = TRUE),
    .groups = "drop"
  )

write_csv(occupation_table, output_path)
write_csv(summary_table, summary_path)

print(occupation_table)
print(summary_table)

cat("\nSaved occupation table to:", output_path, "\n")
cat("Saved occupation summary to:", summary_path, "\n")
